src/block_service/hdfs/hdfs_service.cpp

```diff
@@ -0,0 +1,471 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "hdfs_service.h"
+
+#include <algorithm>
+#include <fstream>
+
+#include <dsn/dist/fmt_logging.h>
+#include <dsn/tool-api/async_calls.h>
+#include <dsn/tool-api/task.h>
+#include <dsn/tool-api/task_tracker.h>
+#include <dsn/utility/error_code.h>
+#include <dsn/utility/filesystem.h>
+#include <dsn/utility/flags.h>
+#include <dsn/utility/safe_strerror_posix.h>
+#include <dsn/utility/utils.h>
+
+namespace dsn {
+namespace dist {
+namespace block_service {
+
+DEFINE_THREAD_POOL_CODE(THREAD_POOL_HDFS_SERVICE)
+DEFINE_TASK_CODE(LPC_HDFS_SERVICE_CALL, TASK_PRIORITY_COMMON, THREAD_POOL_HDFS_SERVICE)
+
+DSN_DEFINE_uint64("replication",
+                  hdfs_read_batch_size_bytes,
+                  64 << 20,
+                  "hdfs read batch size, the default value is 64MB");
+
+DSN_DEFINE_uint64("replication",
+                  hdfs_write_batch_size_bytes,
+                  64 << 20,
+                  "hdfs write batch size, the default value is 64MB");
+
+hdfs_service::hdfs_service() {}
+
+hdfs_service::~hdfs_service()
+{
+    ddebug("Try to disconnect hdfs.");
+    hdfsDisconnect(_fs);
+    _fs = nullptr;
+}
+
+error_code hdfs_service::initialize(const std::vector<std::string> &args)
+{
+    if (args.size() != 2) {
+        return ERR_INVALID_PARAMETERS;
+    }
+    _hdfs_name_node = args[0];
+    _hdfs_path = args[1];
+    _fs = nullptr;
```

Comment:
`_fs` should always be `nullptr`, or you should call `hdfsDisconnect ` before reassign it.